orderbook/src/api.rs

```diff
@@ -41,66 +36,62 @@ pub fn handle_all_routes(
-            .unify()
-            .or(get_fee_and_quote_sell
-                .map(|reply| LabelledReply::new(reply, "get_fee_and_quote_sell")))
-            .unify()
-            .or(get_fee_and_quote_buy
-                .map(|reply| LabelledReply::new(reply, "get_fee_and_quote_buy")))
-            .unify()
-            .or(get_user_orders.map(|reply| LabelledReply::new(reply, "get_user_orders")))
-            .unify(),
+        (create_order.with(handle_metrics("create_order")))
+            .or(get_orders.with(handle_metrics("get_orders")))
+            .or(fee_info.with(handle_metrics("fee_info")))
+            .or(legacy_fee_info.with(handle_metrics("legacy_fee_info")))
+            .or(get_order.with(handle_metrics("get_order")))
+            .or(get_solvable_orders.with(handle_metrics("get_solvable_orders")))
+            .or(get_trades.with(handle_metrics("get_trades")))
+            .or(cancel_order.with(handle_metrics("cancel_order")))
+            .or(get_amount_estimate.with(handle_metrics("get_amount_estimate")))
+            .or(get_fee_and_quote_sell.with(handle_metrics("get_fee_and_quote_sell")))
+            .or(get_fee_and_quote_buy.with(handle_metrics("get_fee_and_quote_buy"))),
     );
-    routes_with_labels
-        .with(wrap_fn(|f| wrap_metrics(f, metrics.clone())))
-        .recover(handle_rejection)
-        .with(cors)
+
+    routes_with_labels.recover(handle_rejection).with(cors)
 }
 
 // We turn Rejection into Reply to workaround warp not setting CORS headers on rejections.
 async fn handle_rejection(err: Rejection) -> Result<impl Reply, Infallible> {
     Ok(err.default_response())
 }
 
-fn wrap_metrics<F>(
-    filter: F,
-    metrics: Arc<Metrics>,
-) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone
-where
-    F: Filter<Extract = (LabelledReply,), Error = Rejection> + Clone + Send + Sync + 'static,
-{
-    warp::any()
-        .and(start_request())
-        .and(filter)
-        .map(move |timer, reply| end_request(metrics.clone(), timer, reply))
+fn handle_metrics(endpoint: impl Into<String>) -> warp::log::Log<impl Fn(warp::log::Info) + Clone> {
+    let metrics = ApiMetrics::instance(get_metric_storage_registry(), endpoint.into()).unwrap();
+
+    warp::log::custom(move |info: warp::log::Info| {
+        metrics
+            .requests_complete
+            .with_label_values(&[info.status().as_str()])
+            .inc();
+        metrics
+            .requests_duration_seconds
+            .observe(info.elapsed().as_secs_f64());
+    })
+}
+
+#[derive(prometheus_metric_storage::MetricStorage, Clone, Debug)]
+#[metric(subsystem = "api", labels("endpoint"))]
+struct ApiMetrics {
+    /// Number of completed API requests.
+    #[metric(labels("status_code"))]
+    requests_complete: prometheus::CounterVec,
+
+    /// Execution time for each API request.
+    #[metric(labels())]
+    requests_duration_seconds: prometheus::Histogram,
 }
```

Comment:
I see that this parameter `metrics: Arc<Metrics>,` was removed from several places, but don't see anything put back to replace it. So at this point I can only assume that this large green block is compensating for all of that. Could you please explain how the code here is instantiated in places where metrics was removed?

Reply:
So we've switched to the global registry that's available through the `get_metric_storage_registry` function. So now we don't have to explicitly pass registry around.